### DISCUSSION TITLE
Two different output categories (how to avoid duplicates)

### DISCUSSION THREAD

**Discussion by oim37** [wintercms/wn-blog-plugin#51] [Q&A] [answered]
Hello, i need help. 
This component is used to display 2 types of posts in 2 main categories (news and knowledge base)

News is displayed via a link `/blog/:slug`
Exmaple uri: /blog/some-new-today-added

The knowledge base is displayed via a different link `/support/kb/:slug`
Exmaple uri: /support/kb/how-update-php-7-to-php-8

In their sections they are filtered in the component settings, each with its own list. And everything seems fine, but if i insert slug from Knowledge base to News it working too and vice versa.
Example uri: /blog/how-update-php-7-to-php-8

and I don’t like it, because when i go to a page in this way, it ends up in Yandex Metrics and Google Analytics and they consider it a new page and after a while they say that it’s a duplicate.

How to avoid this? So that the article related to the news section is available only in /blog/:slug and also with the knowledge base.

kb.htm and blog.htm differ only in links (`url = "/support/kb/:slug"`) and (`url = "/blog/:slug"`)
```
title = "knowledge base"
url = "/support/kb/:slug"
is_hidden = 0

[blogPost]
slug = "{{ :slug }}"
categoryPage = "knowledge-base"
==
<section class="section">
	<div class="container">
	
		<div class="section_header">
<div class="item-kb-block__date">
<i>{{ 'title.publish-change-date'|_ }} {{post.published_at|date('d.m.Y')}}</i></div><br/>
		
			<div class="section_text">			
			</div>
		</div>
		<div class="post">
		{{ post.content_html|raw }}<br/>
		</div>
	</div>
</section>
```

**Comment by mjauvin** [comment, marked as answer]
I don't think there is a setting to avoid this, but you could apply some logic to the `onEnd()` method of each page to avoid showing a post from each other (and maybe redirect to the other page)

```php
function onEnd()
{
    $category = $this->post->categories->first();
    // if ($category && $category->id === 2) {
    if ($category && $category->name === 'news') {
        Return Redirect::to(...);
    }
}
```

**Comment by oim37** [comment]
Good idia, 

Only condition if for some reason it doesn’t work for me:

echo($category);

{"id":2,"name":"\u041d\u043e\u0432\u043e\u0441\u0442\u0438","slug":"news","code":null,"description":"","parent_id":null,"nest_left":1,"nest_right":6,"nest_depth":0,"created_at":"2022-12-17T16:43:19.000000Z","updated_at":"2023-09-20T16:25:50.000000Z","pivot":{"post_id":41,"category_id":2}}

Maybe because of the multilingual. How will it be for "slug" it on all languages same "news"

**Comment by mjauvin** [comment]
Sorry, I should have writen `$category->name === 'news'`

**Comment by mjauvin** [comment]
I updated my earlier code suggestion

**Comment by oim37** [comment]
I undestand, i change name to slug it more better for me and complete link where changing /support/kb/ to /blog/ if article in the news. 

```
function onEnd()
{
   $category = $this->post->categories->first();
    if ($category && $category->slug === 'news') {
        Return Redirect::to(str_replace("/support/kb/", "/blog/", $_SERVER['REQUEST_URI']));
    }
}
```

Is it possible to indicate somewhere 301 redirect in `Return Redirect::to` string?

**Comment by mjauvin** [comment]
The Category id is the only stable property in multi-lingual context (slug and name might differ in different locale for the same category)

**Comment by oim37** [comment]
Yes, I have slug in all languages, always in English. Thank you. 

```
function onEnd()
{
   $category = $this->post->categories->first();
    if ($category && $category->slug === 'news') {
        Return Redirect::to(str_replace("/support/kb/", "/blog/", $_SERVER['REQUEST_URI']), 301);
    }
}
```
This is how it turned out, thank you for your help

**Reply by mjauvin** [reply]
Yes, use `return Redirect::to($url, 301);`

**Comment by mjauvin** [comment]
@oim37 it would be safer to use this to generate the page URL:
```
$url = $this->controller->pageUrl('kb', ['slug' => $this->post->slug], false);
Return Redirect::to($url, 301);
```

Where "kb" is the basefilename of your knowledge page (e.g. support/kb)

**Comment by oim37** [comment]
Okay, that works too, then it will be like this:

```
function onEnd()
{
   $category = $this->post->categories->first();
    if ($category && $category->slug === 'news') {
        $url = $this->controller->pageUrl('kb', ['slug' => $this->post->slug], false);
        Return Redirect::to(str_replace("/support/kb/", "/blog/", $url), 301);
    }
}
```

On the news page i will do a reverse replacement if the article does not apply to it.
Great :)

**Comment by mjauvin** [comment]
Make sure you actually USE `$url` ... in the snippet above it should be this:
```php
function onEnd()
{
   $category = $this->post->categories->first();
    if ($category && $category->slug === 'news') {
        $url = $this->controller->pageUrl('kb', ['slug' => $this->post->slug], false);
        Return Redirect::to($url, 301);
    }
}
```

**Comment by oim37** [comment]
Yes, I didn’t immediately understand all the magic of the code :)

Wonderful
